compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -2537,4 +2563,196 @@ public int hashCode() {
+        // is first value null
+        Iterator iterator = type.valueSpace.iterator();
+        BLangExpression firstElement = (BLangExpression) iterator.next();
+        if ((firstElement == null) || (firstElement.type.getKind() == TypeKind.NIL)) {
+            return true;
+        }
+
+        boolean allMembersHaveSameType = true;
+        boolean defaultFillValuePresent = false;
+
+        // is first value is a valid fill value
+        String defaultFillValue = getDefaultFillValue(firstElement);
+        if (firstElement.toString().equals(defaultFillValue)) {
+            defaultFillValuePresent = true;
+        }
+
+        while (iterator.hasNext()) {
+            Object value =  iterator.next();
+            if (value == null) {
+                return true;
+            }
+
+            BType valueType = ((BLangExpression) value).type;
+            if (valueType.getKind() == TypeKind.NIL) {
+                return true;
+            }
+
+            if (allMembersHaveSameType && !isSameType(valueType, firstElement.type)) {
+                allMembersHaveSameType = false;
+            }
+
+            if (!defaultFillValuePresent && value.toString().equals(defaultFillValue)) {
+                defaultFillValuePresent = true;
+            }
+        }
+
+        if (!allMembersHaveSameType) {
+            return false;
+        }
+
+        return defaultFillValuePresent;
+    }
+
+    private boolean checkFillerValue(BUnionType type) {
+        if (type.isNullable()) {
+            return true;
+        }
+
+        Iterator<BType> iterator = type.getMemberTypes().iterator();
+        BType firstMember = iterator.next();
+        // is first value is a valid fill value
+        if (firstMember.getKind() == TypeKind.NIL) {
+            return true;
+        }
+
+        boolean allMembersHaveSameType = true;
+        boolean defaultFillValuePresent = false;
+
+        // is first value is a valid fill value
+        String defaultFillValue = getDefaultFillValue(firstMember);
+        if (firstMember.toString().equals(defaultFillValue)) {
+            defaultFillValuePresent = true;
+        }
+
+        while (iterator.hasNext()) {
+
+            Object value =  iterator.next();
+            if (value == null) {
+                return true;
+            }
+
+            if (allMembersHaveSameType && !isSameType(firstMember, iterator.next())) {
+                allMembersHaveSameType = false;
+            }
+
+            if (!defaultFillValuePresent && value.toString().equals(defaultFillValue)) {
+                defaultFillValuePresent = true;
+            }
+        }
+
+        if (!allMembersHaveSameType) {
+            return false;
+        }
+        return defaultFillValuePresent;
+    }
+
+    private boolean checkFillerValue(BRecordType type) {
+        for (BField field : type.fields) {
+            if (field.symbol.defaultableParam) {
+                continue;
+            }
+            if (!hasFillerValue(field.type)) {
```

Comment:
IINM, this is not correct. According to the spec, fill member should be done only for lists.

```ballerina
type Foo record {
   string s;
};

Foo[2] arr = [];
```

This should still fail IMO.

Reply:
Commit https://github.com/ballerina-platform/ballerina-lang/pull/20749/commits/f21bca4d1e537732576a0558486343c55dd1def9
should address this.